File: JavaCode/src/com/ysx/leetcode/medium/NumMatchingSubseq.java
package com.ysx.leetcode.medium;

/**
 * Leetcode 792. Number of Matching Subsequences
 * https://leetcode.com/problems/number-of-matching-subsequences/description/
 */
public class NumMatchingSubseq {

    public static void main(String[] args) {
        NumMatchingSubseq test = new NumMatchingSubseq();
        String S = "abcde";
        String[] words = {"a", "bb", "acd", "ace"};
        System.out.println(test.numMatchingSubseq(S, words));
    }

    public int numMatchingSubseq(String S, String[] words) {
        int num = 0;
        int lengthS = S.length();
        for (String word : words) {
            int wordLength = word.length();
            boolean flag = false;
            int i = 0;
            int j = 0;
            while (i < lengthS && j < wordLength) {
                if (S.charAt(i) == word.charAt(j)) {
                    i++;
                    j++;
                    flag = true;
                } else {
                    i++;
                    flag = false;
                }
            }
            if (j == wordLength && flag) {
                num++;
            }
        }
        return num;
    }
}

File: KotlinCode/src/leetcode/medium/NumMatchingSubseq.kt
package leetcode.medium


/**
 * Leetcode 792. Number of Matching Subsequences
 * https://leetcode.com/problems/number-of-matching-subsequences/description/
 */
class NumMatchingSubseq {
    fun numMatchingSubseq(S: String, words: Array<String>): Int {
        var num:Int = 0
        val lengthS:Int = S.length
        for (word:String in words) {
            val lengthWord:Int = word.length
            var flag:Boolean = false
            var i:Int = 0
            var j:Int = 0
            while (i < lengthS && j < lengthWord){
                if (S.get(i) == word.get(j)) {
                    i++
                    j++
                    flag = true
                } else {
                    i++
                    flag = false
                }
            }
            if (j == lengthWord && flag) {
                num++
            }
        }
        return num
    }
}

fun main(args: Array<String>) {
    val test = NumMatchingSubseq()
    val S:String = "abcde"
    val words:Array<String> = arrayOf("a", "bb", "acd", "ace")
    for (word:String in words) {
        println(word)
    }
    print(test.numMatchingSubseq(S, words))
}